File: GlutenFreeAPI/Sources/GlutenFreeAPI/Controllers/FoodController.swift
import Vapor
import Fluent

struct FoodController: RouteCollection {

    func boot(routes: any RoutesBuilder) throws {

        let foods = routes.grouped("foods")

        // GET /foods
        foods.get { req async throws -> Page<Food> in
            let baseQuery = Food.query(on: req.db)
            if let userID = getAuthenticatedUserID(req: req) {
                return try await applyUserSpecificCatalog(req: req, baseQuery: baseQuery, userID: userID)
            }
            return try await baseQuery.paginate(for: req)
        }

        // GET /foods/search?q={query}
        foods.get("search") { req async throws -> Page<Food> in
            guard let searchTerm: String = req.query["q"], !searchTerm.isEmpty else {
                throw Abort(.badRequest, reason: "Missing or empty search query parameter 'q'")
            }

            let searchPattern = "%\(searchTerm)%"
            let baseQuery = Food.query(on: req.db)
                .group(.or) { orGroup in
                    orGroup.filter(.sql(embed: "name ILIKE \(bind: searchPattern)"))
                    orGroup.filter(.sql(embed: "categories ILIKE \(bind: searchPattern)"))
                    orGroup.filter(.sql(embed: "brand ILIKE \(bind: searchPattern)"))
                }

            if let userID = getAuthenticatedUserID(req: req) {
                return try await applyUserSpecificCatalog(req: req, baseQuery: baseQuery, userID: userID)
            }
            return try await baseQuery.paginate(for: req)
        }

        // GET /foods/:id
        foods.get(":foodID") { req async throws -> Food in
            guard let foodIDString = req.parameters.get("foodID"), let foodID = UUID(uuidString: foodIDString) else {
                throw Abort(.badRequest, reason: "Invalid product ID")
            }

            // If user is authenticated, check if this is a custom food or if there is an override
            if let userID = getAuthenticatedUserID(req: req) {
                // Check if it's a custom food first
                if let customOverride = try await UserFoodOverride.query(on: req.db)
                    .filter(\.$user.$id == userID)
                    .filter(\.$id == foodID)
                    .filter(\.$food.$id == nil)
                    .first() {
                    return Food(
                        id: customOverride.id,
                        code: customOverride.code ?? "",
                        name: customOverride.name ?? "",
                        brand: customOverride.brand,
                        categories: customOverride.categories,
                        ingredients: customOverride.ingredients,
                        imageUrl: customOverride.imageUrl,
                        countries: customOverride.countries,
                        glutenFree: customOverride.glutenFree ?? true
                    )
                }

                // Check if it's deleted
                let isDeleted = try await UserFoodOverride.query(on: req.db)
                    .filter(\.$user.$id == userID)
                    .filter(\.$food.$id == foodID)
                    .filter(\.$isDeleted == true)
                    .first() != nil
                
                if isDeleted {
                    throw Abort(.notFound)
                }

                // Check if there is an edit override
                if let override = try await UserFoodOverride.query(on: req.db)
                    .filter(\.$user.$id == userID)
                    .filter(\.$food.$id == foodID)
                    .first() {
                    
                    guard let food = try await Food.find(foodID, on: req.db) else {
                        throw Abort(.notFound)
                    }

                    if let code = override.code { food.code = code }
                    if let name = override.name { food.name = name }
                    if let brand = override.brand { food.brand = brand }
                    if let categories = override.categories { food.categories = categories }
                    if let ingredients = override.ingredients { food.ingredients = ingredients }
                    if let imageUrl = override.imageUrl { food.imageUrl = imageUrl }
                    if let countries = override.countries { food.countries = countries }
                    if let glutenFree = override.glutenFree { food.glutenFree = glutenFree }

                    return food
                }
            }

            // Default base food fetch
            guard let food = try await Food.find(foodID, on: req.db) else {
                throw Abort(.notFound)
            }

            return food
        }

        // Authenticated write routes
        let protectedFoods = foods.grouped(UserMiddleware())

        // POST /foods
        protectedFoods.post(use: createFood)

        // PUT /foods/:foodID
        protectedFoods.put(":foodID", use: updateFood)

        // DELETE /foods/:foodID
        protectedFoods.delete(":foodID", use: deleteFood)
    }

    struct FoodInput: Content {
        var code: String
        var name: String
        var brand: String?
        var categories: String?
        var ingredients: String?
        var imageUrl: String?
        var countries: String?
        var glutenFree: Bool
    }

    // Helper to get authenticated user ID if bearer token is present
    func getAuthenticatedUserID(req: Request) -> UUID? {
        guard let token = req.headers.bearerAuthorization?.token else {
            return nil
        }
        do {
            let payload = try req.jwt.verify(token, as: AuthController.UserPayload.self)
            return UUID(uuidString: payload.subject.value)
        } catch {
            return nil
        }
    }

    // Helper to apply user overrides and append custom foods
    func applyUserSpecificCatalog(
        req: Request,
        baseQuery: QueryBuilder<Food>,
        userID: UUID
    ) async throws -> Page<Food> {
        // 1. Fetch all overrides for this user
        let overrides = try await UserFoodOverride.query(on: req.db)
            .filter(\.$user.$id == userID)
            .all()

        let deletedIDs = Set(overrides.filter { $0.isDeleted }.compactMap { $0.$food.id })
        let editedMap = Dictionary(uniqueKeysWithValues: overrides.filter { !$0.isDeleted && $0.$food.id != nil }.map { ($0.$food.id!, $0) })
        
        let customFoods = overrides.filter { $0.$food.id == nil }.map { override in
            Food(
                id: override.id,
                code: override.code ?? "",
                name: override.name ?? "",
                brand: override.brand,
                categories: override.categories,
                ingredients: override.ingredients,
                imageUrl: override.imageUrl,
                countries: override.countries,
                glutenFree: override.glutenFree ?? true
            )
        }

        // 2. Query all matching base foods excluding deleted ones
        var filteredQuery = baseQuery
        if !deletedIDs.isEmpty {
            filteredQuery = filteredQuery.filter(\.$id !~ deletedIDs)
        }

        // 3. Paginate base foods
        let basePage = try await filteredQuery.paginate(for: req)

        // 4. Apply overrides to base foods in current page
        for food in basePage.items {
            if let foodID = food.id, let override = editedMap[foodID] {
                if let code = override.code { food.code = code }
                if let name = override.name { food.name = name }
                if let brand = override.brand { food.brand = brand }
                if let categories = override.categories { food.categories = categories }
                if let ingredients = override.ingredients { food.ingredients = ingredients }
                if let imageUrl = override.imageUrl { food.imageUrl = imageUrl }
                if let countries = override.countries { food.countries = countries }
                if let glutenFree = override.glutenFree { food.glutenFree = glutenFree }
            }
        }

        // 5. If we are on the first page, prepend the user's custom foods
        var mergedItems = basePage.items
        if basePage.metadata.page == 1 {
            mergedItems.insert(contentsOf: customFoods, at: 0)
        }

        return Page(
            items: mergedItems,
            metadata: PageMetadata(
                page: basePage.metadata.page,
                per: basePage.metadata.per,
                total: basePage.metadata.total + customFoods.count
            )
        )
    }

    // POST /foods
    func createFood(req: Request) async throws -> Response {
        let payload = try req.auth.require(AuthController.UserPayload.self)
        guard let userID = UUID(uuidString: payload.subject.value) else {
            throw Abort(.unauthorized)
        }

        let input = try req.content.decode(FoodInput.self)
        req.logger.info("[VAPOR CRUD] Creating new food for user \(userID). Name: '\(input.name)', Code: '\(input.code)'")

        guard !input.code.isEmpty, !input.name.isEmpty else {
            throw Abort(.badRequest, reason: "Product code and name cannot be empty")
        }

        // Create a private custom food override with food_id = nil
        let override = UserFoodOverride(
            userID: userID,
            foodID: nil,
            isDeleted: false,
            code: input.code,
            name: input.name,
            brand: input.brand,
            categories: input.categories,
            ingredients: input.ingredients,
            imageUrl: input.imageUrl,
            countries: input.countries,
            glutenFree: input.glutenFree
        )

        try await override.save(on: req.db)
        
        // Map to a Food object to return in response
        let food = Food(
            id: override.id,
            code: input.code,
            name: input.name,
            brand: input.brand,
            categories: input.categories,
            ingredients: input.ingredients,
            imageUrl: input.imageUrl,
            countries: input.countries,
            glutenFree: input.glutenFree
        )
        
        req.logger.info("[VAPOR CRUD] Successfully created private custom food with ID: \(override.id?.uuidString ?? "nil")")
        return try await food.encodeResponse(status: .created, for: req)
    }

    // PUT /foods/:foodID
    func updateFood(req: Request) async throws -> Food {
        let payload = try req.auth.require(AuthController.UserPayload.self)
        guard let userID = UUID(uuidString: payload.subject.value) else {
            throw Abort(.unauthorized)
        }

        guard let foodIDString = req.parameters.get("foodID"), let foodID = UUID(uuidString: foodIDString) else {
            throw Abort(.badRequest, reason: "Invalid product ID")
        }

        let input = try req.content.decode(FoodInput.self)
        guard !input.code.isEmpty, !input.name.isEmpty else {
            throw Abort(.badRequest, reason: "Product code and name cannot be empty")
        }

        req.logger.info("[VAPOR CRUD] User \(userID) requested update for food ID: \(foodID) to name: '\(input.name)'")

        // Check if this is a custom private food (food_id is nil, id matches foodID)
        if let existingOverride = try await UserFoodOverride.query(on: req.db)
            .filter(\.$user.$id == userID)
            .filter(\.$id == foodID)
            .filter(\.$food.$id == nil)
            .first() {
            
            existingOverride.code = input.code
            existingOverride.name = input.name
            existingOverride.brand = input.brand
            existingOverride.categories = input.categories
            existingOverride.ingredients = input.ingredients
            existingOverride.imageUrl = input.imageUrl
            existingOverride.countries = input.countries
            existingOverride.glutenFree = input.glutenFree
            
            try await existingOverride.update(on: req.db)
            
            return Food(
                id: existingOverride.id,
                code: input.code,
                name: input.name,
                brand: input.brand,
                categories: input.categories,
                ingredients: input.ingredients,
                imageUrl: input.imageUrl,
                countries: input.countries,
                glutenFree: input.glutenFree
            )
        }

        // Otherwise, this is a base food item. Create or update an override.
        if let existingOverride = try await UserFoodOverride.query(on: req.db)
            .filter(\.$user.$id == userID)
            .filter(\.$food.$id == foodID)
            .first() {
            
            existingOverride.code = input.code
            existingOverride.name = input.name
            existingOverride.brand = input.brand
            existingOverride.categories = input.categories
            existingOverride.ingredients = input.ingredients
            existingOverride.imageUrl = input.imageUrl
            existingOverride.countries = input.countries
            existingOverride.glutenFree = input.glutenFree
            existingOverride.isDeleted = false
            
            try await existingOverride.update(on: req.db)
        } else {
            let newOverride = UserFoodOverride(
                userID: userID,
                foodID: foodID,
                isDeleted: false,
                code: input.code,
                name: input.name,
                brand: input.brand,
                categories: input.categories,
                ingredients: input.ingredients,
                imageUrl: input.imageUrl,
                countries: input.countries,
                glutenFree: input.glutenFree
            )
            try await newOverride.save(on: req.db)
        }

        return Food(
            id: foodID,
            code: input.code,
            name: input.name,
            brand: input.brand,
            categories: input.categories,
            ingredients: input.ingredients,
            imageUrl: input.imageUrl,
            countries: input.countries,
            glutenFree: input.glutenFree
        )
    }

    // DELETE /foods/:foodID
    func deleteFood(req: Request) async throws -> HTTPStatus {
        let payload = try req.auth.require(AuthController.UserPayload.self)
        guard let userID = UUID(uuidString: payload.subject.value) else {
            throw Abort(.unauthorized)
        }

        guard let foodIDString = req.parameters.get("foodID"), let foodID = UUID(uuidString: foodIDString) else {
            throw Abort(.badRequest, reason: "Invalid product ID")
        }

        req.logger.info("[VAPOR CRUD] User \(userID) requested deletion for food ID: \(foodID)")

        // Check if this is a custom private food
        if let existingOverride = try await UserFoodOverride.query(on: req.db)
            .filter(\.$user.$id == userID)
            .filter(\.$id == foodID)
            .filter(\.$food.$id == nil)
            .first() {
            
            try await existingOverride.delete(on: req.db)
            req.logger.info("[VAPOR CRUD] Successfully deleted custom private food \(foodID)")
            return .noContent
        }

        // Otherwise, this is a base food item. Create or update an override setting isDeleted = true.
        if let existingOverride = try await UserFoodOverride.query(on: req.db)
            .filter(\.$user.$id == userID)
            .filter(\.$food.$id == foodID)
            .first() {
            
            existingOverride.isDeleted = true
            try await existingOverride.update(on: req.db)
        } else {
            let newOverride = UserFoodOverride(
                userID: userID,
                foodID: foodID,
                isDeleted: true
            )
            try await newOverride.save(on: req.db)
        }

        req.logger.info("[VAPOR CRUD] Successfully marked base food \(foodID) as deleted for user \(userID)")
        return .noContent
    }
}

File: GlutenFreeKtor/src/main/kotlin/com/glufree/ktor/controllers/FoodController.kt
package com.glufree.ktor.controllers

import com.glufree.ktor.models.*
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.coroutines.Dispatchers
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.SqlExpressionBuilder.isNull
import org.jetbrains.exposed.sql.SqlExpressionBuilder.like
import org.jetbrains.exposed.sql.transactions.experimental.newSuspendedTransaction
import java.time.LocalDateTime
import java.util.UUID

fun Route.foodRoutes() {

    // GET /
    get("/") {
        call.respondText("API Gluten Free funcionando 🚀", contentType = ContentType.Text.Plain)
    }

    route("/foods") {

        // GET /foods
        get {
            val page = call.request.queryParameters["page"]?.toIntOrNull()?.coerceAtLeast(1) ?: 1
            val per = call.request.queryParameters["per"]?.toIntOrNull()?.coerceAtLeast(1) ?: 20

            val limit = per
            val offset = ((page - 1) * per).toLong()

            val userId = call.getAuthenticatedUserId()

            val pageResponse = newSuspendedTransaction(Dispatchers.IO) {
                if (userId != null) {
                    // Fetch all overrides for this user
                    val overrides = UserFoodOverridesTable.selectAll()
                        .where { UserFoodOverridesTable.userId eq userId }
                        .toList()

                    val deletedIDs = overrides.filter { it[UserFoodOverridesTable.isDeleted] }
                        .mapNotNull { it[UserFoodOverridesTable.foodId]?.value }
                        .toSet()

                    val editedMap = overrides.filter { !it[UserFoodOverridesTable.isDeleted] && it[UserFoodOverridesTable.foodId] != null }
                        .associateBy { it[UserFoodOverridesTable.foodId]!!.value }

                    val customFoods = overrides.filter { it[UserFoodOverridesTable.foodId] == null }
                        .map { row ->
                            FoodResponse(
                                id = row[UserFoodOverridesTable.id].value.toString(),
                                code = row[UserFoodOverridesTable.code] ?: "",
                                name = row[UserFoodOverridesTable.name] ?: "",
                                brand = row[UserFoodOverridesTable.brand],
                                categories = row[UserFoodOverridesTable.categories],
                                ingredients = row[UserFoodOverridesTable.ingredients],
                                imageUrl = row[UserFoodOverridesTable.imageUrl],
                                countries = row[UserFoodOverridesTable.countries],
                                glutenFree = row[UserFoodOverridesTable.glutenFree] ?: true,
                                createdAt = row[UserFoodOverridesTable.createdAt]?.toString()
                            )
                        }

                    // Query base foods excluding deleted ones
                    var query = FoodsTable.selectAll()
                    if (deletedIDs.isNotEmpty()) {
                        query = query.where { FoodsTable.id notInList deletedIDs }
                    }

                    val totalCount = query.count()
                    val baseItems = query.limit(limit, offset = offset).map { mapFoodRow(it) }

                    // Apply overrides
                    val updatedItems = baseItems.map { food ->
                        val foodUuid = try { UUID.fromString(food.id) } catch (e: Exception) { null }
                        val override = if (foodUuid != null) editedMap[foodUuid] else null
                        if (override != null) {
                            food.copy(
                                code = override[UserFoodOverridesTable.code] ?: food.code,
                                name = override[UserFoodOverridesTable.name] ?: food.name,
                                brand = override[UserFoodOverridesTable.brand] ?: food.brand,
                                categories = override[UserFoodOverridesTable.categories] ?: food.categories,
                                ingredients = override[UserFoodOverridesTable.ingredients] ?: food.ingredients,
                                imageUrl = override[UserFoodOverridesTable.imageUrl] ?: food.imageUrl,
                                countries = override[UserFoodOverridesTable.countries] ?: food.countries,
                                glutenFree = override[UserFoodOverridesTable.glutenFree] ?: food.glutenFree
                            )
                        } else {
                            food
                        }
                    }

                    var finalItems = updatedItems
                    if (page == 1) {
                        finalItems = customFoods + updatedItems
                    }

                    PageResponse(
                        items = finalItems,
                        metadata = PaginationMetadata(
                            page = page,
                            per = per,
                            total = totalCount + customFoods.size
                        )
                    )
                } else {
                    val totalCount = FoodsTable.selectAll().count()
                    val items = FoodsTable.selectAll().limit(limit, offset = offset).map { mapFoodRow(it) }
                    PageResponse(
                        items = items,
                        metadata = PaginationMetadata(
                            page = page,
                            per = per,
                            total = totalCount
                        )
                    )
                }
            }

            call.respond(pageResponse)
        }

        // GET /foods/search
        get("/search") {
            val searchTerm = call.request.queryParameters["q"]
            if (searchTerm.isNullOrBlank()) {
                call.respond(
                    HttpStatusCode.BadRequest,
                    ErrorResponse(error = true, reason = "Missing or empty search query parameter 'q'")
                )
                return@get
            }

            val page = call.request.queryParameters["page"]?.toIntOrNull()?.coerceAtLeast(1) ?: 1
            val per = call.request.queryParameters["per"]?.toIntOrNull()?.coerceAtLeast(1) ?: 20

            val limit = per
            val offset = ((page - 1) * per).toLong()
            val searchPattern = "%$searchTerm%"

            val userId = call.getAuthenticatedUserId()

            val pageResponse = newSuspendedTransaction(Dispatchers.IO) {
                val searchExpression = (FoodsTable.name.lowerCase() like searchPattern.lowercase()) or
                        (FoodsTable.categories.lowerCase() like searchPattern.lowercase()) or
                        (FoodsTable.brand.lowerCase() like searchPattern.lowercase())

                if (userId != null) {
                    val overrides = UserFoodOverridesTable.selectAll()
                        .where { UserFoodOverridesTable.userId eq userId }
                        .toList()

                    val deletedIDs = overrides.filter { it[UserFoodOverridesTable.isDeleted] }
                        .mapNotNull { it[UserFoodOverridesTable.foodId]?.value }
                        .toSet()

                    val editedMap = overrides.filter { !it[UserFoodOverridesTable.isDeleted] && it[UserFoodOverridesTable.foodId] != null }
                        .associateBy { it[UserFoodOverridesTable.foodId]!!.value }

                    val customFoods = overrides.filter {
                        it[UserFoodOverridesTable.foodId] == null && (
                                (it[UserFoodOverridesTable.name]?.lowercase()?.contains(searchTerm.lowercase()) == true) ||
                                        (it[UserFoodOverridesTable.categories]?.lowercase()?.contains(searchTerm.lowercase()) == true) ||
                                        (it[UserFoodOverridesTable.brand]?.lowercase()?.contains(searchTerm.lowercase()) == true)
                                )
                    }.map { row ->
                        FoodResponse(
                            id = row[UserFoodOverridesTable.id].value.toString(),
                            code = row[UserFoodOverridesTable.code] ?: "",
                            name = row[UserFoodOverridesTable.name] ?: "",
                            brand = row[UserFoodOverridesTable.brand],
                            categories = row[UserFoodOverridesTable.categories],
                            ingredients = row[UserFoodOverridesTable.ingredients],
                            imageUrl = row[UserFoodOverridesTable.imageUrl],
                            countries = row[UserFoodOverridesTable.countries],
                            glutenFree = row[UserFoodOverridesTable.glutenFree] ?: true,
                            createdAt = row[UserFoodOverridesTable.createdAt]?.toString()
                        )
                    }

                    var query = FoodsTable.selectAll().where { searchExpression }
                    if (deletedIDs.isNotEmpty()) {
                        query = query.andWhere { FoodsTable.id notInList deletedIDs }
                    }

                    val totalCount = query.count()
                    val baseItems = query.limit(limit, offset = offset).map { mapFoodRow(it) }

                    val updatedItems = baseItems.map { food ->
                        val foodUuid = try { UUID.fromString(food.id) } catch (e: Exception) { null }
                        val override = if (foodUuid != null) editedMap[foodUuid] else null
                        if (override != null) {
                            food.copy(
                                code = override[UserFoodOverridesTable.code] ?: food.code,
                                name = override[UserFoodOverridesTable.name] ?: food.name,
                                brand = override[UserFoodOverridesTable.brand] ?: food.brand,
                                categories = override[UserFoodOverridesTable.categories] ?: food.categories,
                                ingredients = override[UserFoodOverridesTable.ingredients] ?: food.ingredients,
                                imageUrl = override[UserFoodOverridesTable.imageUrl] ?: food.imageUrl,
                                countries = override[UserFoodOverridesTable.countries] ?: food.countries,
                                glutenFree = override[UserFoodOverridesTable.glutenFree] ?: food.glutenFree
                            )
                        } else {
                            food
                        }
                    }

                    var finalItems = updatedItems
                    if (page == 1) {
                        finalItems = customFoods + updatedItems
                    }

                    PageResponse(
                        items = finalItems,
                        metadata = PaginationMetadata(
                            page = page,
                            per = per,
                            total = totalCount + customFoods.size
                        )
                    )
                } else {
                    val totalCount = FoodsTable.selectAll().where { searchExpression }.count()
                    val items = FoodsTable.selectAll().where { searchExpression }
                        .limit(limit, offset = offset)
                        .map { mapFoodRow(it) }

                    PageResponse(
                        items = items,
                        metadata = PaginationMetadata(
                            page = page,
                            per = per,
                            total = totalCount
                        )
                    )
                }
            }

            call.respond(pageResponse)
        }

        // GET /foods/{foodID}
        get("/{foodID}") {
            val foodIDString = call.parameters["foodID"]
            if (foodIDString.isNullOrBlank()) {
                call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Missing or empty foodID"))
                return@get
            }

            val foodId = try { UUID.fromString(foodIDString) } catch (e: Exception) {
                call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Invalid ID format (expected UUID)"))
                return@get
            }

            val userId = call.getAuthenticatedUserId()

            val response = newSuspendedTransaction(Dispatchers.IO) {
                if (userId != null) {
                    // Check custom private food override with food_id = null and id = foodID
                    val customOverride = UserFoodOverridesTable.selectAll()
                        .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.id eq foodId) and (UserFoodOverridesTable.foodId.isNull()) }
                        .firstOrNull()

                    if (customOverride != null) {
                        return@newSuspendedTransaction HttpStatusCode.OK to FoodResponse(
                            id = customOverride[UserFoodOverridesTable.id].value.toString(),
                            code = customOverride[UserFoodOverridesTable.code] ?: "",
                            name = customOverride[UserFoodOverridesTable.name] ?: "",
                            brand = customOverride[UserFoodOverridesTable.brand],
                            categories = customOverride[UserFoodOverridesTable.categories],
                            ingredients = customOverride[UserFoodOverridesTable.ingredients],
                            imageUrl = customOverride[UserFoodOverridesTable.imageUrl],
                            countries = customOverride[UserFoodOverridesTable.countries],
                            glutenFree = customOverride[UserFoodOverridesTable.glutenFree] ?: true,
                            createdAt = customOverride[UserFoodOverridesTable.createdAt]?.toString()
                        )
                    }

                    // Check if base food is deleted for this user
                    val isDeleted = UserFoodOverridesTable.selectAll()
                        .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) and (UserFoodOverridesTable.isDeleted eq true) }
                        .count() > 0

                    if (isDeleted) {
                        return@newSuspendedTransaction HttpStatusCode.NotFound to ErrorResponse(error = true, reason = "Food not found")
                    }

                    // Check edit override
                    val override = UserFoodOverridesTable.selectAll()
                        .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) }
                        .firstOrNull()

                    val baseFoodRow = FoodsTable.selectAll().where { FoodsTable.id eq foodId }.firstOrNull()
                    if (baseFoodRow == null) {
                        return@newSuspendedTransaction HttpStatusCode.NotFound to ErrorResponse(error = true, reason = "Food not found")
                    }

                    val food = mapFoodRow(baseFoodRow)
                    val finalFood = if (override != null) {
                        food.copy(
                            code = override[UserFoodOverridesTable.code] ?: food.code,
                            name = override[UserFoodOverridesTable.name] ?: food.name,
                            brand = override[UserFoodOverridesTable.brand] ?: food.brand,
                            categories = override[UserFoodOverridesTable.categories] ?: food.categories,
                            ingredients = override[UserFoodOverridesTable.ingredients] ?: food.ingredients,
                            imageUrl = override[UserFoodOverridesTable.imageUrl] ?: food.imageUrl,
                            countries = override[UserFoodOverridesTable.countries] ?: food.countries,
                            glutenFree = override[UserFoodOverridesTable.glutenFree] ?: food.glutenFree
                        )
                    } else {
                        food
                    }

                    HttpStatusCode.OK to finalFood
                } else {
                    val baseFoodRow = FoodsTable.selectAll().where { FoodsTable.id eq foodId }.firstOrNull()
                    if (baseFoodRow != null) {
                        HttpStatusCode.OK to mapFoodRow(baseFoodRow)
                    } else {
                        HttpStatusCode.NotFound to ErrorResponse(error = true, reason = "Food not found")
                    }
                }
            }

            if (response.first == HttpStatusCode.OK) {
                call.respond(response.first, response.second as FoodResponse)
            } else {
                call.respond(response.first, response.second as ErrorResponse)
            }
        }

        // POST /foods
        post {
            call.withAuth { userId ->
                val input = call.receive<FoodInput>()
                if (input.code.isBlank() || input.name.isBlank()) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Product code and name cannot be empty"))
                    return@withAuth
                }

                val newOverrideId = UUID.randomUUID()
                newSuspendedTransaction(Dispatchers.IO) {
                    UserFoodOverridesTable.insert {
                        it[id] = newOverrideId
                        it[UserFoodOverridesTable.userId] = userId
                        it[foodId] = null
                        it[isDeleted] = false
                        it[code] = input.code
                        it[name] = input.name
                        it[brand] = input.brand
                        it[categories] = input.categories
                        it[ingredients] = input.ingredients
                        it[imageUrl] = input.imageUrl
                        it[countries] = input.countries
                        it[glutenFree] = input.glutenFree
                        it[createdAt] = LocalDateTime.now()
                    }
                }

                val food = FoodResponse(
                    id = newOverrideId.toString(),
                    code = input.code,
                    name = input.name,
                    brand = input.brand,
                    categories = input.categories,
                    ingredients = input.ingredients,
                    imageUrl = input.imageUrl,
                    countries = input.countries,
                    glutenFree = input.glutenFree
                )

                call.respond(HttpStatusCode.Created, food)
            }
        }

        // PUT /foods/{foodID}
        put("/{foodID}") {
            call.withAuth { userId ->
                val foodIDString = call.parameters["foodID"]
                if (foodIDString.isNullOrBlank()) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Missing food ID"))
                    return@withAuth
                }

                val foodId = try { UUID.fromString(foodIDString) } catch (e: Exception) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Invalid food ID format"))
                    return@withAuth
                }

                val input = call.receive<FoodInput>()
                if (input.code.isBlank() || input.name.isBlank()) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Product code and name cannot be empty"))
                    return@withAuth
                }

                val updatedFood = newSuspendedTransaction(Dispatchers.IO) {
                    // Check if custom private food
                    val isCustom = UserFoodOverridesTable.selectAll()
                        .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.id eq foodId) and (UserFoodOverridesTable.foodId.isNull()) }
                        .count() > 0

                    if (isCustom) {
                        UserFoodOverridesTable.update({ (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.id eq foodId) and (UserFoodOverridesTable.foodId.isNull()) }) {
                            it[code] = input.code
                            it[name] = input.name
                            it[brand] = input.brand
                            it[categories] = input.categories
                            it[ingredients] = input.ingredients
                            it[imageUrl] = input.imageUrl
                            it[countries] = input.countries
                            it[glutenFree] = input.glutenFree
                        }
                        FoodResponse(
                            id = foodId.toString(),
                            code = input.code,
                            name = input.name,
                            brand = input.brand,
                            categories = input.categories,
                            ingredients = input.ingredients,
                            imageUrl = input.imageUrl,
                            countries = input.countries,
                            glutenFree = input.glutenFree
                        )
                    } else {
                        val baseFoodExists = FoodsTable.selectAll().where { FoodsTable.id eq foodId }.count() > 0
                        if (!baseFoodExists) {
                            return@newSuspendedTransaction null
                        }

                        val existingOverride = UserFoodOverridesTable.selectAll()
                            .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) }
                            .firstOrNull()

                        if (existingOverride != null) {
                            UserFoodOverridesTable.update({ (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) }) {
                                it[code] = input.code
                                it[name] = input.name
                                it[brand] = input.brand
                                it[categories] = input.categories
                                it[ingredients] = input.ingredients
                                it[imageUrl] = input.imageUrl
                                it[countries] = input.countries
                                it[glutenFree] = input.glutenFree
                                it[isDeleted] = false
                            }
                        } else {
                            UserFoodOverridesTable.insert {
                                it[id] = UUID.randomUUID()
                                it[UserFoodOverridesTable.userId] = userId
                                it[this.foodId] = foodId
                                it[isDeleted] = false
                                it[code] = input.code
                                it[name] = input.name
                                it[brand] = input.brand
                                it[categories] = input.categories
                                it[ingredients] = input.ingredients
                                it[imageUrl] = input.imageUrl
                                it[countries] = input.countries
                                it[glutenFree] = input.glutenFree
                                it[createdAt] = LocalDateTime.now()
                            }
                        }

                        FoodResponse(
                            id = foodId.toString(),
                            code = input.code,
                            name = input.name,
                            brand = input.brand,
                            categories = input.categories,
                            ingredients = input.ingredients,
                            imageUrl = input.imageUrl,
                            countries = input.countries,
                            glutenFree = input.glutenFree
                        )
                    }
                }

                if (updatedFood != null) {
                    call.respond(HttpStatusCode.OK, updatedFood)
                } else {
                    call.respond(HttpStatusCode.NotFound, ErrorResponse(error = true, reason = "Food not found"))
                }
            }
        }

        // DELETE /foods/{foodID}
        delete("/{foodID}") {
            call.withAuth { userId ->
                val foodIDString = call.parameters["foodID"]
                if (foodIDString.isNullOrBlank()) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Missing food ID"))
                    return@withAuth
                }

                val foodId = try { UUID.fromString(foodIDString) } catch (e: Exception) {
                    call.respond(HttpStatusCode.BadRequest, ErrorResponse(error = true, reason = "Invalid food ID format"))
                    return@withAuth
                }

                val success = newSuspendedTransaction(Dispatchers.IO) {
                    // Check custom private
                    val isCustom = UserFoodOverridesTable.selectAll()
                        .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.id eq foodId) and (UserFoodOverridesTable.foodId.isNull()) }
                        .count() > 0

                    if (isCustom) {
                        UserFoodOverridesTable.deleteWhere { (UserFoodOverridesTable.userId eq userId) and (id eq foodId) and (UserFoodOverridesTable.foodId.isNull()) }
                        true
                    } else {
                        val baseFoodExists = FoodsTable.selectAll().where { FoodsTable.id eq foodId }.count() > 0
                        if (!baseFoodExists) {
                            return@newSuspendedTransaction false
                        }

                        val existingOverride = UserFoodOverridesTable.selectAll()
                            .where { (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) }
                            .firstOrNull()

                        if (existingOverride != null) {
                            UserFoodOverridesTable.update({ (UserFoodOverridesTable.userId eq userId) and (UserFoodOverridesTable.foodId eq foodId) }) {
                                it[isDeleted] = true
                            }
                        } else {
                            UserFoodOverridesTable.insert {
                                it[id] = UUID.randomUUID()
                                it[UserFoodOverridesTable.userId] = userId
                                it[this.foodId] = foodId
                                it[isDeleted] = true
                                it[createdAt] = LocalDateTime.now()
                            }
                        }
                        true
                    }
                }

                if (success) {
                    call.respond(HttpStatusCode.NoContent)
                } else {
                    call.respond(HttpStatusCode.NotFound, ErrorResponse(error = true, reason = "Food not found"))
                }
            }
        }
    }
}
